plone/formwidget/namedfile/datamanager.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+from plone.formwidget.namedfile.interfaces import IScaleGenerateOnSave
+from plone.namedfile.field import INamedImageField
+from z3c.form.datamanager import AttributeField
+from ZODB.POSException import ConflictError
+from zope.annotation.interfaces import IAnnotations
+from zope.component import adapter
+from zope.component import getMultiAdapter
+from zope.component import getUtility
+from zope.globalrequest import getRequest
+from zope.interface import Interface
+from zope.interface import alsoProvides
+
+import logging
+import os
+import transaction
+
+try:
+    from Products.CMFPlone.factory import _IMREALLYPLONE5  # noqa
+except ImportError:
+    PLONE_5 = False  # pragma: no cover
+else:
+    PLONE_5 = True  # pragma: no cover
+
+
+ANNOTATION_KEY = "plone.formwidget.namedfile.scale"
+ENVIRONMENT_KEY = "PLONE_SCALE_GENERATE_ON_SAVE"
+
+logger = logging.getLogger(__name__)
+
+
+@adapter(Interface, INamedImageField)
+class NamedImageAttributeField(AttributeField):
+
+    def __init__(self, *args, **kwargs):
+        super(NamedImageAttributeField, self).__init__(*args, **kwargs)
+        flag = (os.environ.get(ENVIRONMENT_KEY) or "").lower()
+        self.scale_generate_on_save = flag in ["1", "true", "yes", "on"]
+
+    def set(self, value):
+        """See z3c.form.interfaces.IDataManager"""
+        super(NamedImageAttributeField, self).set(value)
+        if self.scale_generate_on_save:
+            schedule_plone_scale_generate_on_save(
+                self.context, getRequest(), self.field.__name__)
+
+
+def schedule_plone_scale_generate_on_save(context, request, fieldname):
+    annotations = IAnnotations(request, None)
+    if annotations is not None:
+        annotations.setdefault(ANNOTATION_KEY, [])
+        annotations[ANNOTATION_KEY].append((context, fieldname))
+        alsoProvides(request, IScaleGenerateOnSave)
+
+
+def plone_scale_generate_on_save(event):
+    if not IScaleGenerateOnSave.providedBy(event.request):
+        return
+    annotations = IAnnotations(event.request, None)
```

Comment:
I hardly believe that `annotations` can be `None` after we checked `IScaleGenerateOnSave` is provided by event.request.
Anyway, if you define it `annotations = IAnnotations(event.request, {})` you can safely skip the check for it being None.